[Flang][OpenMP] Update declare mapper lookup via use-module

TIFitis opened the pull request:
- Implemented semantic TODO to catch undeclared mappers.
- Fix mapper lookup to include modules imported through USE.
- Update and add tests.

Fixes #163385.

TIFitis committed:
[Flang][OpenMP] Update declare mapper lookup via use-module

	- Implemented semantic TODO to catch undeclared mappers.
	- Fix mapper lookup to include modules imported through USE.
	- Update and add tests.

Fixes #163385.

TIFitis committed:
code cleanup

kiranchandramohan commented:
I don't see changes to add support (in lib/Semantics/mod-file.cpp) for emitting the `declare mapper` directive in the module file (*.mod). How did you manage to get this to work without it?

TIFitis committed:
Emit declare_mapper to .mod file. Update test to check separate module and program file compilation.

TIFitis commented:
@kiranchandramohan Sorry for the earlier misunderstanding — I’m not very familiar with the parsing/semantic side of things.

It was working without exporting to the .mod file because the test originally had both the module and the program in the same file. Once separated, it started failing.

I’ve now updated the PR to export the declare mapper to the .mod file as you suggested, and the lookup works correctly when compiling across different files. I’ve also updated the test accordingly.

As far as I’m aware, FIR/OMP does not emit dummy mappers at any stage.

kiranchandramohan commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -4239,3 +4255,32 @@ void Fortran::lower::genOpenMPRequires(mlir::Operation *mod,
+      }
+    }
+  }
+}
```
Do we still need this function? Can you check how threadprivate works and whether it needs this kind of handling?

TIFitis replied:
Threadprivate directive is serialised into the .mod, and we don't need to lookup symbols by name. We need to lookup existing declare mappers in the module when they are referred to through mapper modifiers, as such, I think we need the `materializeOpenMPDeclareMappers` function.

kiranchandramohan replied:
Do we need to recursively go through all the scopes? Or are only the declare mappers at the top-level needed?

TIFitis replied:
I've reverted to the previous approach.

From what I understand, a function that comes in through a transitive USE is represented in the importing scope by a symbol whose details are `semantics::UseDetails`, pointing back to the defining symbol in the original module. When lowering sees a call, it follows that chain to the ultimate symbol in the source module, so the call is generated exactly as if the function were imported directly. No IR for the callee is emitted in the current compilation unit.

For declare mappers, we need to follow the chain and generate the IR. So we can either take the lazy approach, or we must go through the scopes to find any declare mappers that have brought in through USE modules and lower them.

The declare mappers don't seem to appear at the top-level.

TIFitis committed:
Address reviewer comments. Add test to check declare mapper export to .mod file.

TIFitis committed:
Address copilot comments.

kiranchandramohan reviewed (approved):
I will be away for atleast a week. This needs further review from @agozillon @skatrak or @kparzysz 

Approving to signal that I am not blocking the progress of the patch.

TIFitis commented:
Ping for review.

kparzysz reviewed (approved):
This looks reasonable.  If some parts turn out to be unnecessary, we can remove them later.
